scico/flax/examples.py

```diff
@@ -0,0 +1,1541 @@
+
+    Returns:
+        nd-array of generated data.
+    """
+    if not have_xdesign:
+        raise RuntimeError("Package xdesign is required for use of this function.")
+
+    np.random.seed(seed)
+    saux = np.zeros((ndata, size, size, 1))
+    for i in range(ndata):
+        foam = Foam(size_range=[0.075, 0.0025], gap=1e-3, porosity=1)
+        saux[i, ..., 0] = discrete_phantom(foam, size=size)
+
+    return saux
+
+
+def distributed_data_generation(
+    imgenf: Callable, size: int, nimg: int, sharded: bool = True
+) -> Array:
+    """Data generation distributed among processes.
+
+    Args:
+        imagenf: Function for batch-data generation.
+        size: Size of image to generate.
+        ndata: Number of images to generate.
+        sharded: Flag to indicate if data is to
+            be returned as the chunks generated by
+            each process or consolidated. Default: ``True``.
+
+    Returns:
+        nd-array of generated data.
+    """
+    nproc = jax.device_count()
+    seeds = jnp.arange(nproc)
+    if nproc > 1 and nimg % nproc > 0:
+        raise ValueError("Number of images to generate must be divisible by the number of devices")
+
+    ndata_per_proc = int(nimg // nproc)
+
+    imgs = jax.pmap(imgenf, static_broadcasted_argnums=(1, 2))(seeds, size, ndata_per_proc)
+
+    if not sharded:
+        imgs = imgs.reshape((-1, size, size, 1))
+
+    return imgs
+
+
+def ray_distributed_data_generation(
+    imgenf: Callable, size: int, nimg: int, seedg: float = 123, test_flag: bool = False
+) -> Array:
+    """Data generation distributed among processes using ray.
+
+    Args:
+        imagenf: Function for batch-data generation.
+        size: Size of image to generate.
+        ndata: Number of images to generate.
+        seedg: Base seed for data generation. Default: 123.
+        test_flag: Flag to indicate if running in testing mode. Testing mode requires a different initialization of ray. Default: ``False``.
+
+    Returns:
+        nd-array of generated data.
+    """
+    if not have_ray:
+        raise RuntimeError("Package ray is required for use of this function.")
+
+    if test_flag:
+        ray.init(ignore_reinit_error=True)
+    else:
+        ray.init()
+
+    @ray.remote
+    def data_gen(seed, size, ndata, imgf):
+        return imgf(seed, size, ndata)
+
+    ar = ray.available_resources()
+    # Usage of half available CPU resources.
+    nproc = max(int(ar["CPU"]) // 2, 1)
+    if nproc > nimg:
+        nproc = nimg
+    if nproc > 1 and nimg % nproc > 0:
+        raise ValueError(
+            f"Number of images to generate ({nimg}) must be divisible by the number of available devices ({nproc})"
+        )
+
+    ndata_per_proc = int(nimg // nproc)
+
+    ray_return = ray.get(
+        [data_gen.remote(seed + seedg, size, ndata_per_proc, imgenf) for seed in range(nproc)]
+    )
+    imgs = np.vstack([t for t in ray_return])
+    ray.shutdown()
+
+    return imgs
+
+
+def ct_data_generation(
```

Comment:
Since there is a `load_ct_data`, should this not be `generate_ct_data`?

Reply:
Changed.